l_space/src/main/scala/io/mediachain/Ingress.scala

```
@@ -5,125 +5,169 @@ import Types._
     }
   }
 
-  def defineAuthorship(blobV: Vertex, authorCanonical: Canonical):
-  Xor[AuthorNotFoundError, Unit] = {
-    authorCanonical.vertex(blobV.graph).map { authorCanonicalV =>
-      val existingAuthor = Traversals.getAuthor(blobV.lift).headOption
+  def defineAuthorship(blobV: Vertex, authorCanonicalVertex: Vertex):
+  Xor[GraphError, Unit] = withTransactionXor(blobV.graph) {
 
-      if (!existingAuthor.contains(authorCanonicalV)) {
-        blobV --- AuthoredBy --> authorCanonicalV
+    val authorshipAlreadyDefined = blobV.toPipeline
+      .flatMap(_.findAuthorXor).map { authorCanonical: Canonical =>
+      authorCanonical.canonicalID ==
+        authorCanonicalVertex.toCC[Canonical].canonicalID
+    }.recover {
+      case _: InvalidElementId => {
+        false
       }
+    }
 
-      Xor.right({})
-    }.getOrElse {
-      Xor.left(AuthorNotFoundError(blobV))
+    authorshipAlreadyDefined.map { defined =>
+      if (!defined) {
+        upsertEdge(blobV.graph) {
+          blobV --- AuthoredBy --> authorCanonicalVertex
+        }
+      }
     }
   }
 
-  // throws?
-  def addPerson(graph: Graph,
-                author: Person,
-                raw: Option[RawMetadataBlob] = None):
-  Xor[GraphError, Canonical] = {
-    // If there's an exact match already, return it,
-    // otherwise create a new Person vertex and canonical
-    // and return the canonical
-    val q = Traversals.personBlobsWithExactMatch(graph.V, author)
-
-    val personV: Vertex = q.headOption.getOrElse(graph + author)
-
-    raw.foreach(attachRawMetadata(personV, _))
 
-    graph.V(personV.id)
-      .findCanonicalXor
-      .orElse {
-        val canonicalV = graph + Canonical.create()
-        canonicalV --- DescribedBy --> personV
-        Xor.right(canonicalV.toCC[Canonical])
+  def upsertEdge(graph: Graph)(f: => Edge): Edge = {
+    try f
+    catch {
+      case e: ORecordDuplicatedException => {
+        val id = e.getRid
+        graph.E(id).headOption.getOrElse(
```

Hmm, correction to the above...

Catching the `ORecordDuplicatedException` when you commit the tx does _not_ result in both edges being committed after all.  I was confused by a misleading test.  So I think that's probably the way to go.  It means that upserting edges doesn't need any special handling in the `Ingress` code; it will just silently drop the second edge.  

This could potentially get complicated, if e.g. you add properties to the second edge.  But that wasn't covered by the earlier upsert logic anyway...